stdlib/int16.gr

```diff
@@ -0,0 +1,468 @@
+/**
+ * Utilities for working with the Int16 type.
+ * @example include "int16"
+ *
+ * @since v0.6.0
+ */
+
+module Int16
+
+include "runtime/unsafe/wasmi32"
+include "runtime/exception"
+include "runtime/numbers"
+from Numbers use { tagSimple, coerceNumberToShortInt }
+include "runtime/unsafe/constants"
+from Constants use { _SMIN16_I32, _SMAX16_I32, _SMIN16_I64, _SMAX16_I64 }
+
+@unsafe
+let _DATA_OFFSET = 8n
+@unsafe
+let _TAG_BYTE = 0b10010n
+@unsafe
+let _DATA_MASK = 0xffffff00n
+
+/**
+ * @section Conversions: Functions for converting between Numbers and the Int16 type.
+ */
+
+/**
+ * Converts a Number to an Int16.
+ *
+ * @param number: The value to convert
+ * @returns The Number represented as an Int16
+ *
+ * @since v0.6.0
+ */
+@unsafe
+provide let fromNumber = (number: Number) => {
+  let val = coerceNumberToShortInt(
+    number,
+    _SMIN16_I32,
+    _SMAX16_I32,
+    _SMIN16_I64,
+    _SMAX16_I64,
+    false
+  )
+  let tagged = WasmI32.or(WasmI32.shl(val, _DATA_OFFSET), _TAG_BYTE)
+  WasmI32.toGrain(tagged): Int16
+}
+
+/**
+ * Converts an Int16 to a Number.
+ *
+ * @param value: The value to convert
+ * @returns The Int16 represented as a Number
+ *
+ * @since v0.6.0
+ */
+@unsafe
+provide let toNumber = (value: Int16) => {
+  let num = WasmI32.shrS(WasmI32.fromGrain(value), _DATA_OFFSET)
+  WasmI32.toGrain(tagSimple(num)): Number
+}
+
+@unsafe
+let signExtend = x => WasmI32.shrS(WasmI32.shl(x, _DATA_OFFSET), _DATA_OFFSET)
```

Comment:
Guess what? There's an instruction for this! https://github.com/grain-lang/grain/blob/main/stdlib/runtime/unsafe/wasmi32.gr#L72

Reply:
You'd need to do the operation untagged, but there is indeed an instruction.

Reply:
Yeah I know about that instruction - I even use `extendS16` for int8 as it is valid for tagged int8s. The reason why I didn't use it here is for the reason you mentioned: it only works for untagged values. In this situation it would be more costly to get working on a tagged 16-bit value (right shift 8 + extendS16 + left shift 8 + mask for tag) vs how I do it here (left shift 8 + signed right shift 8)

Reply:
I suppose that's fair. It's only really true for addition and subtraction though, since most of the other operators are fully untagging and retagging, so it's more expensive for things like multiplication, though I imagine that's less common.

Because of that, it might make sense to use the instruction in most places, but for things like addition where we're using tricks we just have a comment about the special tagging.

Reply:
It might also be a good idea to add new primitives for tagging/untagging the small ints, like we have for chars and simple numbers.

Reply:
> Because of that, it might make sense to use the instruction in most places,

Sure thing; although from an inspection I think the extend instructions would only end up being used in multiplication and left shift.

Reply:
Actually, left shift may be better off with how it currently is (meaning only multiplication would change); aside from the "necessary" steps required regardless of implementation (untagging `amount`, `shl` on the values) the current implementation requires an `and`, an `or`, and an `extendS16`, while switching it to untagging/tagging would require an untag (`shr`), an `extendS8`, and a retag (`shl` + `or`); one instruction is saved using the current implementation.